truffle/src/main/java/org/jruby/truffle/language/SnippetNode.java

```diff
@@ -0,0 +1,109 @@
+import org.jruby.truffle.language.methods.DeclarationContext;
+import org.jruby.truffle.language.parser.ParserContext;
+
+public class SnippetNode extends RubyBaseNode {
+
+    private final String expression;
+    private final String[] parameters;
+
+    @CompilationFinal private FrameDescriptor frameDescriptor;
+    @CompilationFinal private FrameSlot[] parameterFrameSlots;
+
+    @Child private DirectCallNode directCallNode;
+
+    public SnippetNode(String expression, String... parameters) {
+        this.expression = expression;
+        this.parameters = parameters;
+    }
+
+    @ExplodeLoop
+    public Object execute(VirtualFrame frame, Object... arguments) {
+        if (directCallNode == null) {
+            CompilerDirectives.transferToInterpreter();
+
+            frameDescriptor = new FrameDescriptor(nil());
+            parameterFrameSlots = new FrameSlot[parameters.length];
+
+            for (int n = 0; n < parameters.length; n++) {
+                parameterFrameSlots[n] = frameDescriptor.findOrAddFrameSlot(parameters[n]);
+            }
+
+            final Source source = Source.fromText(StringOperations.createByteList(expression), "(snippet)");
+
+            final RubyRootNode rootNode = getContext().getCodeLoader().parse(
+                    source,
+                    UTF8Encoding.INSTANCE,
+                    ParserContext.INLINE,
+                    frameDescriptor,
+                    null,
+                    true,
+                    this);
+
+            directCallNode = insert(Truffle.getRuntime().createDirectCallNode(Truffle.getRuntime().createCallTarget(rootNode)));
+
+            if (directCallNode.isInlinable()) {
+                directCallNode.forceInlining();
+            }
+        }
+
+        final Object[] parentFrameArguments = RubyArguments.pack(
+                null,
+                null,
+                RubyArguments.getMethod(frame),
+                DeclarationContext.INSTANCE_EVAL,
+                null,
+                RubyArguments.getSelf(frame),
+                null,
+                new Object[]{});
+
+        final MaterializedFrame parentFrame = Truffle.getRuntime().createMaterializedFrame(
+                parentFrameArguments,
+                frameDescriptor);
+
+        if (arguments.length != parameterFrameSlots.length) {
+            CompilerDirectives.transferToInterpreter();
+            throw new UnsupportedOperationException("number of arguments doesn't match number of parameters");
+        }
+
+        for (int n = 0; n < parameters.length; n++) {
+            parentFrame.setObject(parameterFrameSlots[n], arguments[n]);
+        }
+
+        final Object[] callArguments = RubyArguments.pack(
+                parentFrame,
+                null,
+                RubyArguments.getMethod(frame),
+                DeclarationContext.INSTANCE_EVAL,
+                null,
+                RubyArguments.getSelf(frame),
+                null,
+                new Object[]{});
```

Comment:
Yes but the parser isn't set up to do this at the moment so it will take a little work. I'll look at it separately.